{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61ad40a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7df82a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataset\n",
    "# This data is taken from the 1000 Genome project, on a small subset of positions\n",
    "data=np.load(\"805_SNP_1000G_real.npy\",allow_pickle=True)\n",
    "# You can also download files with more SNPs\n",
    "# Try it later, when you are confident your code is working\n",
    "#data=np.load(\"10K_SNP_1000G_real.npy\",allow_pickle=True)\n",
    "#data=np.load(\"65K_SNP_1000G_real.npy\",allow_pickle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48396477",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look at one element of the dataset\n",
    "print(data[0])\n",
    "# The first coordinate is the super population (continental) ancestry\n",
    "# The second coordinate is the population\n",
    "# The third coordinate is the ID of the individual haplotype, then the list of SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2729300",
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many haplotypes (individuals) are present in the dataset, what is the number of SNPs per individual?\n",
    "number_individual_total = 0\n",
    "number_SNPs = 0\n",
    "#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05abe742",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"There are {number_individual_total} individuals with {number_SNPs} SNPs each.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46ee8f08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We look at the names of super populations in our data set\n",
    "print(set(data[:, 0])) \n",
    "\n",
    "# If you want to know more about the meaning of each ID, you can check:\n",
    "# https://www.internationalgenome.org/data-portal/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af51891",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each super population, compute the number of individual per super population\n",
    "# We initialize the dictionnary which will store the result\n",
    "number_individual_superpop = {'AFR': 0, 'SAS': 0, 'EUR': 0, 'EAS': 0, 'AMR': 0}\n",
    "#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "945cdf8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for pop in number_individual_superpop:\n",
    "    print(f'There are {number_individual_superpop[pop]} individuals in super population {pop}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c44001d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each super population and SNP, compute the allele frequency for that population\n",
    "# We initialize the dictionnary which will store the result\n",
    "allele_frequency_superpop = {'AFR': np.zeros((number_SNPs)), 'SAS': np.zeros((number_SNPs)), 'EUR': np.zeros((number_SNPs)), \\\n",
    "                             'EAS': np.zeros((number_SNPs)), 'AMR': np.zeros((number_SNPs))}\n",
    "# If you are unfamiliar with numpy:  np.zeros((number_SNPs)) can be simply tought and manipulated as a list of size number_SNPs (number of SNPs) initialized with zeros\n",
    "# Fill it with the correct allele frequencies\n",
    "#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6c4fb52",
   "metadata": {},
   "outputs": [],
   "source": [
    "for pop in allele_frequency_superpop:\n",
    "    print(f'The frequency of the first 10 SNPs for super population {pop} is {allele_frequency_superpop[pop][:10]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38be4f78",
   "metadata": {},
   "source": [
    "Compute the FST between each pair of super populations\n",
    "\n",
    "Remember that $FST= \\frac{HT-HS}{HT}$\n",
    "\n",
    "With $HT = 2\\sum_i(f_i(1-f_i))$  with $f_i$ the average allele frequency of allele $i$ for the two populations taken together\n",
    "\n",
    "With $HS = \\sum_i(f_{1,i}(1-f_{1,i}))+(f_{2,i}(1-f_{2,i}))$ with $f_{1,i}$ (resp. $f_{2,i}$ ) the average allele frequency of allele $i$ for population 1 (resp. 2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2405c155",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_HT(allele_frequency):\n",
    "    # The result will be a dic of dic as we compute the allele frequency for each pair of super populations\n",
    "    # For example HT['AFR']['EUR'] will contain the HT value for these two populations taken together\n",
    "    HT = {}\n",
    "    # TODO\n",
    "    return HT \n",
    "\n",
    "def compute_HS(allele_frequency):\n",
    "    HS = {}\n",
    "    # TODO\n",
    "    return HS\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f6042c",
   "metadata": {},
   "outputs": [],
   "source": [
    "HT = compute_HT(allele_frequency_superpop)\n",
    "HS = compute_HS(allele_frequency_superpop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbd150f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We compute the FST from HT and HS\n",
    "def compute_FST(HT,HS):\n",
    "    FST = {}\n",
    "    # TODO\n",
    "    return FST\n",
    "\n",
    "# TODO uncomment the next line...\n",
    "# FST = compute_FST(HT,HS)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4c524db",
   "metadata": {},
   "source": [
    "We plot the FST matrix, what do you notice? Give an interpretation of the results.\n",
    "\n",
    "Note: If you do not have zeros on diagonal, there is a mistake!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9557f83c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_fst(FST,cell_size=1, map = {}):\n",
    "    if map == {}:\n",
    "        keys = sorted(FST.keys())\n",
    "    else:\n",
    "        keys = sorted(map, key=map.get)\n",
    "\n",
    "\n",
    "    plt.figure(figsize=(cell_size*len(keys), cell_size*len(keys)))\n",
    "    # Create a 2D numpy array\n",
    "    matrix = np.array([[round(FST[row][col],2) for col in keys] for row in keys])\n",
    "\n",
    "    # Plot heatmap\n",
    "    plt.imshow(matrix, cmap=\"viridis\")\n",
    "    #plt.colorbar()\n",
    "\n",
    "    # Set ticks and labels\n",
    "    plt.xticks(range(len(keys)), keys)\n",
    "    plt.yticks(range(len(keys)), keys)\n",
    "\n",
    "    # Optionally add values on cells\n",
    "    for i in range(len(keys)):\n",
    "        for j in range(len(keys)):\n",
    "            plt.text(j, i, str(matrix[i, j]), ha='center', va='center', color='white')\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "    \n",
    "# TODO uncomment the next line...\n",
    "#plot_fst(FST)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd926b07",
   "metadata": {},
   "source": [
    "We now want to try FST at population level, re-use your previous fonctions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00bb10d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "populations = set(data[:, 1])\n",
    "print(f' The populations are {populations}')\n",
    "number_individual_pop ={pop: 0 for pop in populations}\n",
    "# Count the number of individual per population\n",
    "#TODO\n",
    "allele_frequency_pop = {pop: np.zeros((number_SNPs)) for pop in populations}\n",
    "# Fill it with the correct allele frequencies\n",
    "#TODO\n",
    "HT = compute_HT(allele_frequency_pop)\n",
    "HS = compute_HS(allele_frequency_pop)\n",
    "FST = compute_FST(HT,HS)\n",
    "\n",
    "# We simply map each population to its super population so that we plot populations of the same super population next to each others\n",
    "mapping = {second: first for first, second in data[:, :2]}\n",
    "\n",
    "plot_fst(FST,cell_size=0.5,map=mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "854c4bb8",
   "metadata": {},
   "source": [
    "What do you observe? How can you interpret it"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7e7171",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lplancheenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
